runtime/src/main/cpp/Memory.cpp

```diff
@@ -2359,13 +2272,20 @@ OBJ_GETTER(ReadHeapRefLocked, ObjHeader** location, int32_t* spinlock) {
   MEMORY_LOG("ReadHeapRefLocked: %p\n", location)
   lock(spinlock);
   ObjHeader* value = *location;
-  // We do not use UpdateRef() here to avoid having ReleaseRef() on return slot under the lock.
-  if (value != nullptr) {
-    AddStackRef(value);
-  }
   unlock(spinlock);
-  updateReturnRefAdded(OBJ_RESULT, value);
-  return value;
+  if (value != nullptr)
+    rememberNewContainer(value->container());
```

Comment:
It's too late to do this after `unlock`.
Another thread may have changed `*location` and performed GC by this moment.
`IncrementRC` must be done under lock.